dashboard/app/controllers/registrations_controller.rb

```diff
@@ -120,13 +120,27 @@ def set_email
     return head(:bad_request) if params[:user].nil?
 
     successfully_updated =
-      if forbidden_change?(current_user, params)
-        false
-      elsif needs_password?(current_user, params)
-        current_user.update_with_password(set_email_params)
+      if current_user.migrated?
+        if forbidden_change?(current_user, params)
+          false
+        elsif needs_password?(current_user, params)
+          if current_user.valid_password?(params[:user][:current_password])
+            current_user.update_primary_authentication_option(params[:user][:email])
+          else
+            false
+          end
+        else
+          current_user.update_primary_authentication_option(params[:user][:email])
```

Comment:
This still needs to change.

When the user is under 13, `params[:user][:email]` will _never_ be populated, but they can still "change their email" by submitting a new hashed email to our server.

Here's the code on the client that omits email if the user is under 13.

https://github.com/code-dot-org/code-dot-org/blob/a1adf33ec94e7cb73a8be9b9eea3388f8888308a/apps/src/lib/ui/accounts/ChangeEmailController.js#L118-L119

I think our practice so far has been for the server to accept email, hashed_email, or both.
- If email is provided, the server generates a hashed email and saves both for a teacher, or just the hash for a student.
- If hashed email is provided, this is an error for a teacher, but can be used to update a student.
- If both are provided I think we trust the cleartext and the server regenerates the hash? Not sure though.

We can make some new choices about what we allow here, but we need to support the scenario where the client only sends a hash.

While we're at it, should we be using strong parameters here, or do we not need it since this is a fairly narrow case?

Reply:
oh, i see. i thought that the email would always be submitted as plaintext. i'll update the endpoint and `update_primary_authentication_option` to account for a hashed email from the front-end.

we don't need strong params for this since it's already implemented:
https://github.com/code-dot-org/code-dot-org/blob/17e6a693350ce6ff841e834b574fee4408c15299/dashboard/app/controllers/registrations_controller.rb#L258-L263

Reply:
We're not using that implementation though; notice that `set_email_params` is called in the non-migrated case below (new lines 139, 142) but not in the new migrated case, above.

Reply:
oh, sorry about that. i was thinking of the `authenticate_scope!` before action, which i believe authenticates using strong params.

since the old endpoint was ultimately using the [devise method](https://github.com/code-dot-org/code-dot-org/blob/17e6a693350ce6ff841e834b574fee4408c15299/dashboard/app/controllers/registrations_controller.rb#L258-L263) `update_with_password`, it made more sense to me to use strong params, whereas the new `update_primary_authentication_option` method can't do that since it requires more manual work and can't use the devise method.

Reply:
although, since we define `set_email_params` and most likely won't be using `update_primary_authentication_method` elsewhere, maybe it does make more sense to still use strong params